website/versioned_docs/version-v1.x/circuits.md

```diff
@@ -7,45 +7,408 @@ sidebar_position: 7
 
 # Circuits
 
-MACI has three zk-SNARK circuits:
+MACI has three zk-SNARK [circuits](https://github.com/privacy-scaling-explorations/maci/tree/dev/circuits):
 
-1. `ProcessMessages.circom`, which takes a batch of messages, and updates the
+1. [`ProcessMessages.circom`](https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/circom/processMessages.circom), which takes a batch of messages, and updates the
```

Comment:
> which takes a batch of messages, and updates the state and ballot trees according to the contents of said messages.

This confuses me - how does a circuit "update the state and ballot trees"? It's not actually updating state anywhere is it? How can a circuit alter state elsewhere?

Reply:
I see how this can be misleading, I will rephrase! basically I think what they meant to say there is that the processMessages circuit will try to decrypt the messages, and based on the content of the message, update within itself the trees, to generate a proof that the processing was done correctly offchain! In other words, the circuit takes a final state, an initial state, and the leaves (messages and user signups) - it process these messages via the different state transitions to finally check that the expected state is correct

Reply:
> I see how this can be misleading, I will rephrase! basically I think what they meant to say there is that the processMessages circuit will try to decrypt the messages, and based on the content of the message, update within itself the trees, to generate a proof that the processing was done correctly offchain! In other words, the circuit takes a final state, an initial state, and the leaves (messages and user signups) - it process these messages via the different state transitions to finally check that the expected state is correct

Ah, great explanation, thanks!